File: daq/slc/javalib/b2rc/src/b2rc/core/TTDNode.java
package b2rc.core;

public class TTDNode extends RCNode {

	static public final String TAG = "ttd_node";

	private FTSW[] _ftsw_v = new FTSW[16];
	private int _ftsw_i = 0;

	public TTDNode(String name) {
		super(name);
	}

	public TTDNode() {
	}

	public FTSW getFTSW(int i) {
		return _ftsw_v[i];
	}

	public FTSW[] getFTSWs() {
		return _ftsw_v;
	}

	public void addFTSW(FTSW ftsw) {
		if (_ftsw_i < _ftsw_v.length) {
			_ftsw_v[_ftsw_i] = ftsw;
			_ftsw_i++;
		}
	}

	public void clearFTSWs() {
		for (int i = 0; i < _ftsw_v.length; i++)
			_ftsw_v[i] = null;
		_ftsw_i = 0;
	}

	@Override
	public String getSQLFields() {
		StringBuffer buf = new StringBuffer();
		buf.append(super.getSQLFields());
		for (int i = 0; i < _ftsw_v.length; i++) {
			buf.append(", ftsw_id_" + i + " smallint");
		}
		return buf.toString();
	}

	@Override
	public String getSQLLabels() {
		StringBuffer buf = new StringBuffer();
		buf.append(super.getSQLLabels());
		for (int i = 0; i < _ftsw_v.length; i++) {
			buf.append(", ftsw_id_" + i);
		}
		return buf.toString();
	}

	@Override
	public String getSQLValues() {
		StringBuffer buf = new StringBuffer();
		buf.append(super.getSQLValues());
		for (int i = 0; i < _ftsw_v.length; i++) {
			if (_ftsw_v[i] != null) {
				buf.append(", " + _ftsw_v[i].getId());
			} else {
				buf.append(", -1");
			}
		}
		return buf.toString();
	}

	@Override
	public String getTag() {
		return TAG;
	}

}

File: daq/slc/src/base/src/TTDNode.cc
#include "TTDNode.h"

#include "FTSW.h"

#include "StringUtil.h"

#include <sstream>

using namespace Belle2;

void TTDNode::clearFTSWs() throw()
{
  for (size_t i = 0; i < MAX_FTSWS; i++)
    _ftsw_v[i] = NULL;
  _ftsw_i = 0;
}

const std::string TTDNode::getSQLFields() const throw()
{
  std::stringstream ss;
  ss << NSMNode::getSQLFields();
  for (size_t i = 0; i < MAX_FTSWS; i++) {
    ss << ", ftsw_id_" << i << " smallint";
  }
  return ss.str();
}

const std::string TTDNode::getSQLLabels() const throw()
{
  std::stringstream ss;
  ss << NSMNode::getSQLLabels();
  for (size_t i = 0; i < MAX_FTSWS; i++) {
    ss << ", ftsw_id_" << i;
  }
  return ss.str();
}

const std::string TTDNode::getSQLValues() const throw()
{
  std::stringstream ss;
  ss << NSMNode::getSQLValues();
  for (size_t i = 0; i < MAX_FTSWS; i++) {
    int id = (i < _ftsw_i && _ftsw_v[i] != NULL) ? _ftsw_v[i]->getID() : -1;
    ss << ", " << id;
  }
  return ss.str();
}

int TTDNode::getParams(const Command& command, unsigned int* pars,
                       std::string& datap)
{
  int npar = 0;
  std::stringstream ss; ss.str("");
  if (command == Command::BOOT) {
    pars[npar++] = _ftsw_i;
    pars[npar++] = 0;
    for (size_t i = 0; i < _ftsw_i; i++) {
      if (_ftsw_v[i] != NULL) {
        pars[1] |= _ftsw_v[i]->isUsed() << i;
        pars[npar++] =  _ftsw_v[i]->getChannel();
        ss << _ftsw_v[i]->getFirmware() << " ";
      }
    }
  } else if (command == Command::LOAD) {
  } else if (command == Command::TRIGFT) {
    for (size_t i = 0; i < _ftsw_i; i++) {
      if (_ftsw_v[i] != NULL) {
        pars[npar++] = _ftsw_v[i]->getTriggerMode();
        pars[npar++] = _ftsw_v[i]->getDummyRate();
        pars[npar++] = _ftsw_v[i]->getTriggerLimit();
      }
    }
  }
  datap = ss.str();
  return npar;
}

void TTDNode::setParams(const Command& command, int npar,
                        const unsigned int* pars,
                        const std::string& datap)
{
  int par_i = 0;
  if (command == Command::BOOT) {
    _ftsw_i = pars[par_i++];
    par_i++;
    std::vector<std::string> str_v = Belle2::split(datap, ' ');
    for (size_t i = 0; i < _ftsw_i; i++) {
      if (_ftsw_v[i] == NULL) {
        _ftsw_v[i] = new FTSW();
      }
      _ftsw_v[i]->setUsed(0x01 & (pars[1] >> i));
      _ftsw_v[i]->setChannel(pars[par_i++]);
      _ftsw_v[i]->setFirmware(str_v[i]);
    }
  } else if (command == Command::LOAD) {
  } else if (command == Command::TRIGFT) {
    for (size_t i = 0; i < _ftsw_i; i++) {
      if (_ftsw_v[i] != NULL) {
        _ftsw_v[i]->setTriggerMode(pars[par_i++]);
        _ftsw_v[i]->setDummyRate(pars[par_i++]);
        _ftsw_v[i]->setTriggerLimit(pars[par_i++]);
      }
    }
  }
}

